snuba/subscriptions/verifier.py

```diff
@@ -99,3 +112,203 @@ def close(self, timeout: Optional[float] = None) -> None:
+                )
+            ),
+            override_topics=[orig_result_topic, new_result_topic],
+        ),
+        # We have to pass a single topic here to keep the stream processor happy. But it's
+        # ignored since we passed both of our topics to override_topics to the result consumer.
+        orig_result_topic,
+        VerifierProcessingFactory(
+            orig_result_topic=orig_result_topic,
+            new_result_topic=new_result_topic,
+            threshold_sec=threshold_sec,
+            metrics=metrics,
+        ),
+    )
+
+
+class ResultTopic(Enum):
+    ORIGINAL = "original"
+    NEW = "new"
+
+
+class ResultStore:
+    """
+    Keeps track of messages received on the old and new result topics grouped by
+    their scheduled subscription time.
+    """
+
+    def __init__(self, threshold_sec: int, metrics: MetricsBackend) -> None:
+        # Maps timestamp to the count of messages received within each second
+        self.__stores: Mapping[ResultTopic, MutableMapping[int, int]] = {
+            ResultTopic.ORIGINAL: {},
+            ResultTopic.NEW: {},
+        }
+
+        # The last timestamp we recorded metrics for
+        # (or the time we will start recording)
+        self.__timestamp_low_watermark: Optional[int] = None
+
+        # The most recent timestamp we've seen
+        self.__timestamp_high_watermark: Optional[int] = None
+
+        self.__threshold_sec = threshold_sec
+        self.__metrics = metrics
+
+    def increment(self, topic: ResultTopic, item: SubscriptionResultData) -> None:
+        # When we get the first message, set the low watermark
+        if self.__timestamp_low_watermark is None:
+            self.__timestamp_low_watermark = item.timestamp + self.__threshold_sec
+
+        # Advance the high watermark if needed
+        if (
+            self.__timestamp_high_watermark is None
+            or item.timestamp > self.__timestamp_high_watermark
+        ):
+            self.__timestamp_high_watermark = item.timestamp
+
+        # Ensure the message is not below the low watermark
+        if item.timestamp <= self.__timestamp_low_watermark:
+            # Record `stale_message` unless the consumer is just starting up
+            if (
+                self.__timestamp_high_watermark
+                >= self.__timestamp_low_watermark + self.__threshold_sec
+            ):
+                self.__metrics.increment("stale_message")
+            return
+
+        # Increment counters
+        store = self.__stores[topic]
+
+        if item.timestamp not in store:
+            store[item.timestamp] = 0
+        store[item.timestamp] += 1
+
+        # Record metrics and advance the low watermark
+        low_watermark = self.__timestamp_high_watermark - self.__threshold_sec
+
+        if low_watermark > self.__timestamp_low_watermark:
+            self._record(low_watermark)
+            self.__timestamp_low_watermark = low_watermark
+
+    def _record(self, until: int) -> None:
```

Comment:
I don't think that can happen since the StreamProcessor which manages calling `submit` doesn't use any threading. I don't think we are going to be doing any threading in the verifier processing in general as there is not really IO tasks to to do.